strawberryfields/io.py

```diff
@@ -146,6 +173,63 @@ def to_program(bb):
     return prog
 
 
+def _convert_tdm(bb):
+    # pylint: disable=import-outside-toplevel
+    from strawberryfields.tdm.tdmprogram import TDMProgram
+
+    prog = TDMProgram(max(bb.modes) + 1, name=bb.name)
+
+    args = [bb._var[f"p{i}"].flatten() for i in range(max(bb.modes) + 1)]
+    # append the quantum operations
+    with prog.context(*args, copies=bb.programtype["options"]["copies"]) as (p, q):
+        for op in bb.operations:
```

Comment:
Thought about this as well, and decided to go for duplication instead for two reasons:

1. It made it clearer to understand what was happening, avoiding the use of a lot of `if-tdmprogram` branches.

2. Things might change with TDMProgram in the future, and this might make it easier to maintain both program conversions, allowing for more program-specific changes in the future. It's difficult to say what might change, but since they use completely different contexts and also differ in the use of symbolic parameters (including the use of rolling/unrolling in `TDMProgram`) there could be significant updates here that would make it easier to maintain if separating the two conversions.

I'm happy to make this more compact though if others think that's better.

Reply:
I would keep this separate for now as well; it is still too early, and we might make changes in the near term that cause the logic to deviate